### ISSUE TITLE
Generalize Matrix class

### ISSUE THREAD

**Issue by mdymczyk** [h2oai/h2o4gpu#259]
The Matrix class in tSVD is a great start and would be a good idea to use it across the project. For that, we'd need to make it multi-GPU, where all the data is equally distributed across GPUs. Distribution strategy should be: shuffled, "as-is".

**Comment by mdymczyk** [comment]
@navdeep-G if you're ever feeling bored :-)

**Comment by navdeep-G** [comment]
@mdymczyk Is the idea here to use a single matrix class across algos? Basically a single representation of data?

**Comment by mdymczyk** [comment]
Yes, instead of coding the same thing multiple times. Or at least to have a common base Matrix class and then specialize it for each algorithm as necessary. But this is only for when you have spare time and nothing with higher priority to do.

**Comment by pseudotensor** [comment]
We discussed this a bit before, but I don't recall... .What about the pogs matrix class?

**Comment by mdymczyk** [comment]
@pseudotensor that would be ideal, since then all the algos could use the same code. Also, we'd get (from what I remember) sparse matrices.

Now that I'm thinking about it maybe making the matrix class itself multi-GPU isn't that good of an idea? Since the underlying operations might be too algo specific. I.e. when POGS is doing A * B it does so on the whole matrices on each GPU (since the data is replicated fully) but in most cases in KMeans I just want to multiply a part of matrix A on GPUX against part of matrix B on GPUX.

**Comment by pseudotensor** [comment]
Pogs is currently agnostic to multiGPU.  The matrix classes are generic, except the class stores which GPU the data is on.  If one wanted to extend it to multi-GPU in some way, say equal partitions of the data, that could be a fairly simple extension.
